optional-goi-analysis/goi-calculations.R

```diff
@@ -0,0 +1,301 @@
+
+# Source in goi functions 
+source(file.path(project_root, "utils", "provided-goi-analysis-functions.R"))
+
+# Check R and Bioconductor versions
+check_r_bioc_versions()
+
+# Load project
+setup_renv(project_filepath = project_root)
+
+## Set the seed
+set.seed(opt$seed)
+
+# Check that the input file exists
+if (!file.exists(opt$sce)){
+  stop(paste(opt$sce, "does not exist."))
+}
+
+# Check that the outout directory exists
+if (!dir.exists(opt$output_directory)) {
+  dir.create(opt$output_directory, recursive = TRUE)
+}
+
+#### Load libraries ------------------------------------------------------------
+
+suppressPackageStartupMessages({
+  library(magrittr)
+  library(readr)
+  library(AnnotationDbi)
+  library(org.Hs.eg.db)
+  library(ComplexHeatmap)
+  library(SingleCellExperiment)
+})
+
+#### Read in data --------------------------------------------------------------
+
+goi_list <- data.table::fread(opt$input_goi_list, stringsAsFactors = FALSE)
+sce <- read_rds(file.path(opt$sce))
+
+#### Perform mapping -----------------------------------------------------------
+
+if (opt$perform_mapping == TRUE) {
+  # turn column name into symbol for pulling the column info out of data frame
+  identifier_column_name <- rlang::sym(opt$identifier_column_name)
+  ids_for_mapping <- goi_list %>%
+    dplyr::pull(identifier_column_name)
+  
+  # Define the annotation packages based on the specified organism
+  annotation_list <- list(
+    'Homo sapiens' = "org.Hs.eg.db",
+    'Mus musculus' = "org.Mm.eg.db",
+    'Danio rerio' = "org.Dr.eg.db"  #,
+    # new organisms would go here
+  )
+  # Error handling
+  if (!(opt$organism %in% names(annotation_list))) {
+    stop(paste(opt$organism, "is not supported"))
+  }
+  # Load library and assign annotation package
+  library(annotation_list[[opt$organism]], character.only = TRUE)
+  annotation_package <-
+    eval(parse(text = annotation_list[[opt$organism]]))
+  
+  # Perform mapping
+  goi_list <- mapIds(
+    # organism annotation package
+    annotation_package,
+    # the provided gene identifiers
+    keys = ids_for_mapping,
+    # the type of provided gene identifiers
+    keytype = opt$input_identifiers,
+    # the type of gene identifiers to map to
+    column = opt$output_identifiers,
+    multiVals = opt$multi_mappings
+  ) %>%
+    tibble::enframe(
+      name = opt$identifier_column_name,
+      value = tolower(opt$output_identifiers)
+    ) %>%
+    # enframe() makes a `list` column; we will simplify it with unnest()
+    # This will result in one row of our data frame per list item
+    tidyr::unnest(cols = tolower(opt$output_identifiers)) %>%
+    # grab only the unique rows
+    dplyr::distinct() %>%
+    # join the remaining columns
+    dplyr::left_join(goi_list, by = tolower(opt$identifier_column_name))
+  
+  # Save mapped object to file
+  write_tsv(goi_list, file.path(
+    opt$output_directory,
+    paste0(opt$library_id, "_mapped_genes.tsv")
+  ))
+}
```

Comment:
```suggestion
  ))
  
  # define mapped goi associated with rownmaes of SCE object 
  goi_rownames_column <- tolower(opt$sce_rownames_identifier)
  goi_rownames <- goi_list %>%
    dplyr::pull(goi_rownames_column)
    
} else {

  # if no mapping is performed then set goi to input id column 
  goi_rownames <- goi_list %>%
    dplyr::pull(gene_id)
  }
```

Here I would set the variables for grabbing the genes from the SCE object based on if mapping was performed or not. I also used my suggestion from earlier on setting the input ID column to always be `gene_id` and removing that argument entirely.